plugins/pulp_docker/plugins/registry.py

```diff
@@ -194,34 +231,125 @@ def create_download_request(self, image_id, file_name, destination_dir):
         self.add_auth_header(req)
         return req
 
-    def add_auth_header(self, request):
+
+class V2Repository(object):
+    """
+    This class represents a Docker v2 repository.
+    """
+    API_VERSION_CHECK_PATH = '/v2/'
+    LAYER_PATH = '/v2/{name}/blobs/{tarsum}'
+    MANIFEST_PATH = '/v2/{name}/manifests/{reference}'
+    TAGS_PATH = '/v2/{name}/tags/list'
+
+    def __init__(self, name, download_config, registry_url, working_dir):
         """
-        Given a download request, add an Authorization header if we have an
-        auth token available.
+        Initialize the V2Repository.
 
-        :param request: a download request
-        :type  request: nectar.request.DownloadRequest
+        :param name:            name of a docker repository
+        :type  name:            basestring
+        :param download_config: download configuration object
+        :type  download_config: nectar.config.DownloaderConfig
+        :param registry_url:    URL for the docker registry
+        :type  registry_url:    basestring
+        :param working_dir:     full path to the directory where files should
+                                be saved
+        :type  working_dir:     basestring
         """
-        if self.token:
-            if request.headers is None:
-                request.headers = {}
-            # this emulates what docker itself does
-            request.headers['Authorization'] = 'Token %s' % self.token
+        self.name = name
+        self.download_config = download_config
+        self.registry_url = registry_url
+        self.downloader = HTTPThreadedDownloader(self.download_config, AggregatingEventListener())
+        self.working_dir = working_dir
 
-    def get_image_url(self):
+    def api_version_check(self):
         """
-        Get a URL for the registry or the endpoint, for use in retrieving image
-        files. The "endpoint" is a host name that might be returned in a header
-        when retrieving repository data above.
+        Make a call to the registry URL's /v2/ API call to determine if the registry supports API
+        v2. If it does not, raise NotImplementedError. If it does, return.
+        """
+        _logger.debug('Determining if the registry URL can do v2 of the Docker API.')
+        exception = NotImplementedError('%s is not a Docker v2 registry.' % self.registry_url)
 
-        :return:    a url that is either the provided registry url, or if an
-                    endpoint is known, that same url with the host replaced by
-                    the endpoint
-        :rtype:     basestring
+        try:
+            headers, body = self._get_path(self.API_VERSION_CHECK_PATH)
+        except IOError:
+            raise exception
+
+        try:
+            version = headers['Docker-Distribution-API-Version']
+        except KeyError:
+            raise exception
+
+        if version != "registry/2.0":
```

Comment:
The header value can specify multiple versions. The values are passed the normal way when multiple values are specified for a single header, but I'm not sure how the client code here would represent that.